### PR TITLE
LUCENE-9317: Clean up split package in analyzers-common

### PR DESCRIPTION
**Pull request by mocobeta**
This is a draft pull request for review, to try to clean up package name conflicts between `analyzers-common` and `core`. Also I tried to make necessary changes as small as possible.

See https://issues.apache.org/jira/browse/LUCENE-9317 for more background.

The main changes are:

- Move analysis base classes to lucene-core (o.a.l.a) from analyzers-common (o.a.l.a.util)
- Rename all service provider files (META-INF/services/...).
- Move o.a.l.a.standard.StandardTokenizer to lucene-core
- Split o.a.l.a.standard in analyzers-common into o.a.l.a.classic and o.a.l.a.email

With above changes, there is no package name conflicts.

- `o.a.l.a.util` and newly created `o.a.l.a.classic` and `o.a.l.a.email` only exist in `analyzers-common`
- `o.a.l.a.standard` only exists in `lucene-core`
- other packages are not touched.

Compiling whole Lucene/Solr main classes are fine, thanks to IDE's refactoring feature.

Tasks to be done:

- Create fake factory base classes in o.a.l.a.util for backward compatibility (?)
- Fix tests
- Fix gradle scripts (?)

### CODE REVIEW

**Comment by mocobeta** [comment]
To add to the volume of the changes here, the most of them are automatically done by IDE.

**Comment by mocobeta** [comment]
This branch has been a bit stale (I just merged recent master and resolved some conflicts).
@uschindler do you have time to review this, or would it be better that I close it for the moment?

**Comment by dweiss** [comment]
In general, I like it. I wondered whether it'd make sense to create a separate package (subproject) for those classes moved to the core - this would make the core really separate from analysis. If somebody is using dependency systems with transitive dependencies, they wouldn't feel the difference (the analysis-util/ analysis-base?) subprojects would be sucked in automatically. What do you think?

**Comment by mocobeta** [comment]
It should be able to create a subproject for classes I moved to core here. 

> this would make the core really separate from analysis.

I don't fully understand that... we cannot completely remove `o.a.l.analysis` from lucene-core in any case (e.g.,  `o.a.l.document.Field` uses `o.a.l.a.Analysis`).

**Comment by dweiss** [comment]
Ah, sorry. Ok then, Tomoko!

**Comment by mocobeta** [comment]
@uschindler seems busy.

I don't want to maintain this branch for very long (the diff is so large), but I need at least one reviewer to proceed this.
@dweiss would you take care this, if you have some time?

**Comment by dweiss** [comment]
Hi Tomoko. The patch looks good to me (precommit doesn't pass though). I would commit it in once you get precommit to work - this issue has been out there for a while, nobody objected. If there is a need for changes (on master), we'll just follow-up.

**Review by uschindler — CHANGES_REQUESTED** [review]
Hi @mocobeta, 
sorry for being late. This looks like my original proposal: Move the factory classes to core, I have no problem with that!

I have not yet checked if the regeneration of jflex files for the classic tokenizer work with gradle, but it's more important to move the stuff. Fixing up the build can be done later.

In a second step we may further move some utility classes away from analysis to general lucene utility classes (org.pacahe.lucene.util). ResourceLoader is in my opinion one candidate.
The SPI stuff looks fine now, be sure to also move the META-INF files, because you renamed classes but did not change the files inside META-INF/services.

We now also need to fix the tests (and duplicate them in all modules) that checks if all analyzers have a META-INF/services entry.

If you have any questions, please ask me on slack, there I can respond faster.

**Comment by uschindler** [comment]
> Create fake factory base classes in o.a.l.a.util for backward compatibility (?)

We do this only in Lucene 9, so more important to add all changes to MIGRATE.md

> Fix tests

I mentioned this, as the META-INF/services files are not updated. This makes renamed analyzers not load, as SPI can't find them
As said before we need an SPI load test that ensures that all analyzer coponents have a factory that loads successfully with SPI. Maybe move that test (abstract) to test-framework and create a test implementation instance for each module containing factories. The test in analysis/common is not enough anymore.

> Fix gradle scripts (?)

jflex regenerate may need to be adapted.

**Comment by mocobeta** [comment]
Thanks @dweiss and @uschindler, I will make the necessary changes and let you know then.

> precommit doesn't pass though

Yes, because many tests need to be fixed (I changed only main classes for early stage review).

**Comment by dweiss** [comment]
If there are test infrastructure that needs to be shared then I'd suggest creating a project that is a test-configuration dependency from other subprojects (rather than cloning those classes). This is a simple and cheap thing to do with gradle.

**Comment by mocobeta** [comment]
> If you have any questions, please ask me on slack, there I can respond faster.

We already have Jira and Github, I'd rather not disperse discussions onto multiple platforms any more...
Is Jira/Github mentions insufficient for us?

**Comment by msokolov** [comment]
+1 for that - slack is opaque and not as good for asynchronous comms IMO.
However it does serve to get someone's attention sometimes - "Hey Uwe! go
read your email!" or something. I do find email a challenge though. There
are so many messages here it's very difficult to keep up and find the
important ones

**Comment by mocobeta** [comment]
I'm using Slack for my daily work and other open/close communities since it's the first place for them, but I'd like to limit interactions with other lucene/solr committers on Jira/Github... It's tiring to hustle to catch someone with multiple channels (slack or twitter or...).

If I fail to catch attentions of right people on Jira/Github or dev list for issues, I'll simply take it that I'm ignored because it's not important thing for them.

**Comment by dweiss** [comment]
Jira or github is fine. A direct e-mail is fine too if you need to bring somebody's attention to something in particular. Don't take it personally, Tomoko - you need to take into account that people leave on holidays, have breaks and generally may work on open source when they have time (or enthusiasm), to do so. This isn't ignoring you, it's life getting in the way of hobbies. ;)

**Comment by mocobeta** [comment]
I don't take it personally - open source is also a hobby for me so I don't want to make extra efforts to catch attentions of others... there are no difference between jira/github mentions and private mail, slack mention or sns direct messages.

**Comment by mocobeta** [comment]
My wording could be bad? I'm not yet good at English to communicate my feelings or thoughts correctly :)

I don't take it badly that someone to not to respond or delay to reply (I tend to be slow to respond too), at the same time I think we don't have no obligation to try hard to get feedback from others.

**Comment by mocobeta** [comment]
> I have not yet checked if the regeneration of jflex files for the classic tokenizer work with gradle, but it's more important to move the stuff. Fixing up the build can be done later.

Gradle script for jflex was fixed in https://github.com/apache/lucene-solr/pull/1836/commits/6782fcf21d1a3e08607c8c7f22ef712c37eef8ab. 
`./gradlew jflex` now works for me.

> In a second step we may further move some utility classes away from analysis to general lucene utility classes (org.pacahe.lucene.util). ResourceLoader is in my opinion one candidate.

ResourceLoader and ResourceLoaderAware were moved to o.a.l.a.util: https://github.com/apache/lucene-solr/pull/1836/commits/f3db4f5ef7bf576d876eaf810c2b70d3a8e9e9b4

> The SPI stuff looks fine now, be sure to also move the META-INF files, because you renamed classes but did not change the files inside META-INF/services.

All SPI service files in META-INF/services (lucene/ and solr/) were renamed.

> We now also need to fix the tests (and duplicate them in all modules) that checks if all analyzers have a META-INF/services entry.

> I mentioned this, as the META-INF/services files are not updated. This makes renamed analyzers not load, as SPI can't find them
As said before we need an SPI load test that ensures that all analyzer coponents have a factory that loads successfully with SPI. Maybe move that test (abstract) to test-framework and create a test implementation instance for each module containing factories. The test in analysis/common is not enough anymore.

I refactored all unit tests for both of lucene and solr, also cleaned up them so that there is no duplicated tests.

> We do this only in Lucene 9, so more important to add all changes to MIGRATE.md

I added an entry to MIGRATE.md: https://github.com/apache/lucene-solr/pull/1836/commits/2f6f8118e815ec2b2bd3b29a8314c2427464b50d


**All tests under lucene/ is OK and this branch now passes `precomit`** so I think most of the work was already done. There is one thing that I need help to make this ready to the master - `org.apache.solr.pkg.TestPackages` fails with NoClassDefFoundError:

```
$ ./gradlew :solr:core:test --tests "org.apache.solr.pkg.TestPackages" -Ptests.jvms=6 -Ptests.jvmargs=-XX:TieredStopAtLevel=1

 2> Caused by: java.lang.NoClassDefFoundError: org/apache/lucene/analysis/util/CharFilterFactory
  2>    at java.base/java.lang.ClassLoader.defineClass1(Native Method)
  2>    at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1016)
  2>    at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
  2>    at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:550)
  2>    at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:458)
  2>    at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:452)
  2>    at java.base/java.security.AccessController.doPrivileged(Native Method)
  2>    at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:451)
  2>    at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
  2>    at java.base/java.net.FactoryURLClassLoader.loadClass(URLClassLoader.java:899)
  2>    at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
  2>    at java.base/java.lang.Class.forName0(Native Method)
  2>    at java.base/java.lang.Class.forName(Class.java:398)
  2>    at org.apache.solr.core.SolrResourceLoader.findClass(SolrResourceLoader.java:500)
  2>    at org.apache.solr.core.SolrResourceLoader.newInstance(SolrResourceLoader.java:574)
  2>    at org.apache.solr.pkg.PackageListeningClassLoader.newInstance(PackageListeningClassLoader.java:107)
  2>    at org.apache.solr.schema.FieldTypePluginLoader$1.create(FieldTypePluginLoader.java:271)
  2>    at org.apache.solr.schema.FieldTypePluginLoader$1.create(FieldTypePluginLoader.java:254)
  2>    at org.apache.solr.util.plugin.AbstractPluginLoader.load(AbstractPluginLoader.java:157)
  2>    at org.apache.solr.schema.FieldTypePluginLoader.readAnalyzer(FieldTypePluginLoader.java:295)
  2>    at org.apache.solr.schema.FieldTypePluginLoader.create(FieldTypePluginLoader.java:100)
  2>    at org.apache.solr.schema.FieldTypePluginLoader.create(FieldTypePluginLoader.java:49)
```

I tried to figure out out where the package or class name is hard coded but failed to do so.

**Comment by mocobeta** [comment]
@uschindler I hope this will catch your attention and the timing is good for you. :)

I don't have enough time and energy neither to keep it up-to-date with the master for long, nor chase very busy people to hurry them for review.

**Comment by uschindler** [comment]
Hi,

I am in hospital for an urgency with appendix. I can review stuff, but I only have my Chromebook available, so limited programming capacity.

Hope to get back to full availability soon.

Uwe 

Am September 27, 2020 10:14:29 AM UTC schrieb Tomoko Uchida <notifications@github.com>:
>@uschindler I hope this will catch your attention and the timing is
>good for you. :)
>
>I don't have enough time and energy neither to keep it up-to-date with
>the master for long, nor chase very busy people to hurry them for
>review.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/lucene-solr/pull/1836#issuecomment-699615254

--
Uwe Schindler
Achterdiek 19, 28357 Bremen
https://www.thetaphi.de

**Comment by mocobeta** [comment]
> I am in hospital for an urgency with appendix. I can review stuff, but I only have my Chromebook available, so limited programming capacity.

What happened, please take care of yourself before considering this stuff (it's just a code)...

**Comment by dweiss** [comment]
Ugh. Take care of yourself (and your appendix), Uwe. I'll review this later today.

**Comment by uschindler** [comment]
Thanks Dawid,

The appendix (not my wife, the part of intestine) is gone already.  Unfortunately it broke before it was extracted so I have to stay a bit longer.

Uwe

Am September 27, 2020 2:14:47 PM UTC schrieb Dawid Weiss <notifications@github.com>:
>Ugh. Take care of yourself (and your appendix), Uwe. I'll review this
>later today.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/lucene-solr/pull/1836#issuecomment-699640851

--
Uwe Schindler
Achterdiek 19, 28357 Bremen
https://www.thetaphi.de

**Comment by dweiss** [comment]
Hope you'll recover soon, Uwe!

**Comment by dweiss** [comment]
That test has hardcoded classes (schema-plugins.jar.bin) referencing things before the refactoring - hence the ClassNotFoundException. I think it's a poor test considering future maintenance (in that it contains binary classes). @noblepaul Is there any way to change it somehow?

**Review by dweiss — APPROVED** [review]
I've looked at the changes and diffed them against master. They look good to me. The migration entry is good. I'd add an entry to CHANGES.txt too (both for Solr and Lucene maybe).

As for the offending Solr test: I would file an issue and add an AwaitsFix to it so that it can be corrected as a separate issue. Those binary classes are not very convenient anyway so perhaps there is a different way of implementing that test (like compiling that ZIP on the fly from classes available in test classpath).

File: lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamFactoryTestCase.java
Comment (mocobeta) [inline lucene/test-framework/src/java/org/apache/lucene/analysis/BaseTokenStreamFactoryTestCase.java:43]: This is finally moved to test-framework from analyzers-common. :)

**Comment by mocobeta** [comment]
Thanks @dweiss for review.

>  I'd add an entry to CHANGES.txt too (both for Solr and Lucene maybe).

I added a changes entry for Lucene https://github.com/apache/lucene-solr/pull/1836/commits/a31632d17703803b23883323ea974d24842759e5. (There is no Solr issue, though its internal was refactored.)

> I would file an issue and add an AwaitsFix to it so that it can be corrected as a separate issue. Those binary classes are not very convenient anyway so perhaps there is a different way of implementing that test (like compiling that ZIP on the fly from classes available in test classpath).

I opened https://issues.apache.org/jira/browse/SOLR-14901 and referred it from `@AwaitsFix` https://github.com/apache/lucene-solr/pull/1836/commits/2fdeb7ff5cded3fe9f9a7aaaa36bfd42e997fd17, then this branch now passes `./gradlew -p solr/ tests`.
I'm not sure the title or description of SOLR-14901 is appropriate. Can you please fix it if needed?

I'll merge it to master after running whole tests again.

**Comment by dweiss** [comment]
I think this is fine, Tomoko. I'll take care of the Solr test issue in a spare moment.

**Comment by mocobeta** [comment]
I will keep open https://issues.apache.org/jira/browse/LUCENE-9317. 
@uschindler if you notice any follow-ups are needed, please give comments there.

**Comment by dweiss** [comment]
I ran a full check, passes.

**Comment by mocobeta** [comment]
> I ran a full check, passes.

Thank you!

**Pull request merged**
